File: android/app/src/main/kotlin/com/strawberrytiger/game/rendering/TigerRenderer.kt
package com.strawberrytiger.game.rendering

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path
import android.graphics.RectF
import com.strawberrytiger.game.GameState
import com.strawberrytiger.game.Tiger
import kotlin.math.sin

class TigerRenderer {

    private val paint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val path = Path()
    private val ovalRect = RectF()

    private fun drawRotatedOval(canvas: Canvas, cx: Float, cy: Float, rx: Float, ry: Float, rotationRad: Float, p: Paint) {
        canvas.save()
        canvas.translate(cx, cy)
        canvas.rotate(Math.toDegrees(rotationRad.toDouble()).toFloat())
        ovalRect.set(-rx, -ry, rx, ry)
        canvas.drawOval(ovalRect, p)
        canvas.restore()
    }

    private fun drawOvalAt(canvas: Canvas, cx: Float, cy: Float, rx: Float, ry: Float, p: Paint) {
        ovalRect.set(cx - rx, cy - ry, cx + rx, cy + ry)
        canvas.drawOval(ovalRect, p)
    }

    // Port of roundRect helper
    private fun drawRoundRect(canvas: Canvas, x: Float, y: Float, w: Float, h: Float, radius: Float, p: Paint) {
        ovalRect.set(x, y, x + w, y + h)
        canvas.drawRoundRect(ovalRect, radius, radius, p)
    }

    fun draw(canvas: Canvas, tiger: Tiger, frameCount: Int, gameState: GameState) {
        val x = tiger.x
        val y = tiger.y

        // --- Tail ---
        paint.style = Paint.Style.STROKE
        paint.strokeCap = Paint.Cap.ROUND

        val tipWag = sin(frameCount * 0.12f) * 4f

        // Orange tail
        paint.color = Color.parseColor("#FF8C00")
        paint.strokeWidth = 5f
        path.reset()
        path.moveTo(x + 8f, y + 20f)
        path.quadTo(x - 6f, y + 16f, x - 14f, y + 12f + tipWag)
        canvas.drawPath(path, paint)

        // Black rings on tail
        paint.color = Color.parseColor("#1a1a1a")
        paint.strokeWidth = 2f
        canvas.drawLine(x - 1f, y + 17.5f, x - 3f, y + 19f, paint)
        canvas.drawLine(x - 7f, y + 15f, x - 9f, y + 16.5f, paint)
        canvas.drawLine(x - 12f, y + 12.5f + tipWag * 0.7f, x - 14f, y + 14f + tipWag * 0.7f, paint)

        // --- Body ---
        paint.style = Paint.Style.FILL
        paint.color = Color.parseColor("#FF8C00")
        drawRoundRect(canvas, x + 6f, y + 10f, 46f, 32f, 8f, paint)

        // White belly
        paint.color = Color.parseColor("#FFF3E0")
        drawOvalAt(canvas, x + 30f, y + 36f, 18f, 6f, paint)

        // Body stripes
        paint.style = Paint.Style.STROKE
        paint.color = Color.parseColor("#1a1a1a")
        paint.strokeWidth = 2.5f
        // Stripe 1
        path.reset()
        path.moveTo(x + 16f, y + 12f)
        path.quadTo(x + 14f, y + 24f, x + 17f, y + 36f)
        canvas.drawPath(path, paint)
        // Stripe 2
        path.reset()
        path.moveTo(x + 24f, y + 11f)
        path.quadTo(x + 22f, y + 22f, x + 25f, y + 35f)
        canvas.drawPath(path, paint)
        // Stripe 3
        path.reset()
        path.moveTo(x + 32f, y + 11f)
        path.quadTo(x + 34f, y + 22f, x + 31f, y + 36f)
        canvas.drawPath(path, paint)
        // Stripe 4
        path.reset()
        path.moveTo(x + 40f, y + 12f)
        path.quadTo(x + 42f, y + 24f, x + 39f, y + 35f)
        canvas.drawPath(path, paint)

        // --- Ears ---
        paint.style = Paint.Style.FILL
        // Left ear
        paint.color = Color.parseColor("#FF8C00")
        drawRotatedOval(canvas, x + 43f, y + 3f, 5f, 6f, -0.15f, paint)
        // Right ear
        drawRotatedOval(canvas, x + 57f, y + 3f, 5f, 6f, 0.15f, paint)

        // Black ear backs (only top half: from PI to 2*PI)
        // We approximate by drawing a smaller oval shifted up
        paint.color = Color.parseColor("#1a1a1a")
        // Left ear back - top half
        canvas.save()
        canvas.translate(x + 43f, y + 1f)
        canvas.rotate(Math.toDegrees(-0.15).toFloat())
        canvas.clipRect(-5f, -5f, 5f, 0f)
        ovalRect.set(-4f, -4f, 4f, 4f)
        canvas.drawOval(ovalRect, paint)
        canvas.restore()
        // Right ear back - top half
        canvas.save()
        canvas.translate(x + 57f, y + 1f)
        canvas.rotate(Math.toDegrees(0.15).toFloat())
        canvas.clipRect(-5f, -5f, 5f, 0f)
        ovalRect.set(-4f, -4f, 4f, 4f)
        canvas.drawOval(ovalRect, paint)
        canvas.restore()

        // White ear spots
        paint.color = Color.WHITE
        canvas.drawCircle(x + 43f, y + 1f, 2f, paint)
        canvas.drawCircle(x + 57f, y + 1f, 2f, paint)

        // --- Head ---
        paint.color = Color.parseColor("#FFA500")
        drawOvalAt(canvas, x + 50f, y + 13f, 13f, 12f, paint)

        // White muzzle
        paint.color = Color.parseColor("#FFF8E1")
        drawOvalAt(canvas, x + 50f, y + 18f, 8f, 6f, paint)

        // White cheek patches
        paint.color = Color.parseColor("#FFF3E0")
        drawRotatedOval(canvas, x + 42f, y + 15f, 4f, 3f, -0.3f, paint)
        drawRotatedOval(canvas, x + 58f, y + 15f, 4f, 3f, 0.3f, paint)

        // Face stripes
        paint.style = Paint.Style.STROKE
        paint.color = Color.parseColor("#1a1a1a")
        paint.strokeWidth = 2f
        // Left face stripes
        path.reset()
        path.moveTo(x + 40f, y + 6f)
        path.quadTo(x + 42f, y + 9f, x + 44f, y + 8f)
        canvas.drawPath(path, paint)
        path.reset()
        path.moveTo(x + 38f, y + 9f)
        path.quadTo(x + 41f, y + 12f, x + 43f, y + 11f)
        canvas.drawPath(path, paint)
        // Right face stripes
        path.reset()
        path.moveTo(x + 60f, y + 6f)
        path.quadTo(x + 58f, y + 9f, x + 56f, y + 8f)
        canvas.drawPath(path, paint)
        path.reset()
        path.moveTo(x + 62f, y + 9f)
        path.quadTo(x + 59f, y + 12f, x + 57f, y + 11f)
        canvas.drawPath(path, paint)

        // --- Eyes ---
        paint.style = Paint.Style.FILL
        // White
        paint.color = Color.WHITE
        drawRotatedOval(canvas, x + 45f, y + 11f, 3.5f, 2.5f, -0.1f, paint)
        drawRotatedOval(canvas, x + 55f, y + 11f, 3.5f, 2.5f, 0.1f, paint)
        // Amber iris
        paint.color = Color.parseColor("#E6A800")
        canvas.drawCircle(x + 46f, y + 11f, 2f, paint)
        canvas.drawCircle(x + 56f, y + 11f, 2f, paint)
        // Black pupils
        paint.color = Color.BLACK
        canvas.drawCircle(x + 46f, y + 11f, 1f, paint)
        canvas.drawCircle(x + 56f, y + 11f, 1f, paint)

        // --- Nose ---
        paint.color = Color.parseColor("#D4576B")
        drawOvalAt(canvas, x + 50f, y + 16f, 3f, 2f, paint)
        // Black half (bottom half only)
        paint.color = Color.parseColor("#1a1a1a")
        canvas.save()
        canvas.translate(x + 50f, y + 15.5f)
        canvas.clipRect(-3f, 0f, 3f, 2f)
        ovalRect.set(-2.5f, -1.5f, 2.5f, 1.5f)
        canvas.drawOval(ovalRect, paint)
        canvas.restore()

        // --- Mouth ---
        paint.style = Paint.Style.STROKE
        paint.color = Color.parseColor("#1a1a1a")
        paint.strokeWidth = 1f
        canvas.drawLine(x + 50f, y + 17.5f, x + 50f, y + 19f, paint)
        // Left arc
        ovalRect.set(x + 44f, y + 16f, x + 50f, y + 22f)
        canvas.drawArc(ovalRect, 0f, 180f, false, paint)
        // Right arc
        ovalRect.set(x + 50f, y + 16f, x + 56f, y + 22f)
        canvas.drawArc(ovalRect, 0f, 180f, false, paint)

        // --- Whiskers ---
        paint.color = Color.WHITE
        paint.strokeWidth = 1.5f
        paint.strokeCap = Paint.Cap.ROUND
        // Left whiskers
        canvas.drawLine(x + 43f, y + 17f, x + 32f, y + 15f, paint)
        canvas.drawLine(x + 43f, y + 18f, x + 31f, y + 19f, paint)
        canvas.drawLine(x + 43f, y + 19f, x + 33f, y + 23f, paint)
        // Right whiskers
        canvas.drawLine(x + 57f, y + 17f, x + 68f, y + 15f, paint)
        canvas.drawLine(x + 57f, y + 18f, x + 69f, y + 19f, paint)
        canvas.drawLine(x + 57f, y + 19f, x + 67f, y + 23f, paint)

        // --- Legs ---
        paint.color = Color.parseColor("#FF8C00")
        paint.strokeWidth = 4f
        paint.strokeCap = Paint.Cap.ROUND

        if (tiger.isOnGround && gameState == GameState.PLAYING) {
            val legPhase = sin(frameCount * 0.3f)
            // Front legs
            canvas.drawLine(x + 38f, y + 40f, x + 38f + legPhase * 6f, y + 50f, paint)
            canvas.drawLine(x + 44f, y + 40f, x + 44f - legPhase * 6f, y + 50f, paint)
            // Back legs
            canvas.drawLine(x + 16f, y + 40f, x + 16f - legPhase * 6f, y + 50f, paint)
            canvas.drawLine(x + 22f, y + 40f, x + 22f + legPhase * 6f, y + 50f, paint)
        } else if (!tiger.isOnGround) {
            // Tucked legs in air
            canvas.drawLine(x + 38f, y + 40f, x + 42f, y + 46f, paint)
            canvas.drawLine(x + 44f, y + 40f, x + 48f, y + 46f, paint)
            canvas.drawLine(x + 16f, y + 40f, x + 12f, y + 46f, paint)
            canvas.drawLine(x + 22f, y + 40f, x + 18f, y + 46f, paint)
        } else {
            // Standing still (menu)
            canvas.drawLine(x + 38f, y + 40f, x + 38f, y + 50f, paint)
            canvas.drawLine(x + 44f, y + 40f, x + 44f, y + 50f, paint)
            canvas.drawLine(x + 16f, y + 40f, x + 16f, y + 50f, paint)
            canvas.drawLine(x + 22f, y + 40f, x + 22f, y + 50f, paint)
        }
    }
}

File: ios/Sources/StrawberryTiger/Rendering/TigerRenderer.swift
import SpriteKit
import UIKit

enum TigerRenderer {

    /// The tiger drawing area: origin at (0,0) in the texture corresponds to
    /// the top-left corner of the bounding box that is 14px left of tiger.x and 1px above tiger.y.
    /// Texture size: 90 x 55 (covers tail tip through right whisker, ears through legs).
    static let textureWidth: CGFloat = 90
    static let textureHeight: CGFloat = 55

    /// Offset from tiger.x/tiger.y to the texture origin (top-left in game coords).
    /// The tail extends 14px left of tiger.x, ears ~1px above tiger.y.
    static let originOffsetX: CGFloat = 14
    static let originOffsetY: CGFloat = 1

    static func makeTexture(
        isOnGround: Bool,
        isPlaying: Bool,
        frameCount: Int
    ) -> SKTexture {
        let w = textureWidth
        let h = textureHeight
        let renderer = UIGraphicsImageRenderer(size: CGSize(width: w, height: h))
        let image = renderer.image { uiCtx in
            let ctx = uiCtx.cgContext

            // The local coordinate origin (lx, ly) maps to what game.js calls (tiger.x, tiger.y).
            let lx: CGFloat = originOffsetX
            let ly: CGFloat = originOffsetY

            // Helper: draw rounded rect into current path
            func addRoundedRect(_ rect: CGRect, radius: CGFloat) {
                let path = UIBezierPath(roundedRect: rect, cornerRadius: radius)
                ctx.addPath(path.cgPath)
            }

            ctx.setLineCap(.round)

            // --- Tail ---
            let tipWag = sin(CGFloat(frameCount) * 0.12) * 4

            ctx.setStrokeColor(UIColor(red: 1, green: 0.549, blue: 0, alpha: 1).cgColor) // #FF8C00
            ctx.setLineWidth(5)
            ctx.beginPath()
            ctx.move(to: CGPoint(x: lx + 8, y: ly + 20))
            ctx.addQuadCurve(to: CGPoint(x: lx - 14, y: ly + 12 + tipWag),
                             control: CGPoint(x: lx - 6, y: ly + 16))
            ctx.strokePath()

            // Tail rings
            ctx.setStrokeColor(UIColor(red: 0.102, green: 0.102, blue: 0.102, alpha: 1).cgColor) // #1a1a1a
            ctx.setLineWidth(2)
            ctx.beginPath()
            ctx.move(to: CGPoint(x: lx - 1, y: ly + 17.5))
            ctx.addLine(to: CGPoint(x: lx - 3, y: ly + 19))
            ctx.strokePath()
            ctx.beginPath()
            ctx.move(to: CGPoint(x: lx - 7, y: ly + 15))
            ctx.addLine(to: CGPoint(x: lx - 9, y: ly + 16.5))
            ctx.strokePath()
            ctx.beginPath()
            ctx.move(to: CGPoint(x: lx - 12, y: ly + 12.5 + tipWag * 0.7))
            ctx.addLine(to: CGPoint(x: lx - 14, y: ly + 14 + tipWag * 0.7))
            ctx.strokePath()

            // --- Body ---
            ctx.setFillColor(UIColor(red: 1, green: 0.549, blue: 0, alpha: 1).cgColor)
            addRoundedRect(CGRect(x: lx + 6, y: ly + 10, width: 46, height: 32), radius: 8)
            ctx.fillPath()

            // White belly
            ctx.setFillColor(UIColor(red: 1, green: 0.953, blue: 0.878, alpha: 1).cgColor) // #FFF3E0
            let bellyPath = UIBezierPath(ovalIn: CGRect(
                x: lx + 30 - 18, y: ly + 36 - 6, width: 36, height: 12))
            ctx.addPath(bellyPath.cgPath)
            ctx.fillPath()

            // Body stripes
            ctx.setStrokeColor(UIColor(red: 0.102, green: 0.102, blue: 0.102, alpha: 1).cgColor)
            ctx.setLineWidth(2.5)

            let stripes: [(CGPoint, CGPoint, CGPoint)] = [
                (CGPoint(x: lx + 16, y: ly + 12), CGPoint(x: lx + 14, y: ly + 24), CGPoint(x: lx + 17, y: ly + 36)),
                (CGPoint(x: lx + 24, y: ly + 11), CGPoint(x: lx + 22, y: ly + 22), CGPoint(x: lx + 25, y: ly + 35)),
                (CGPoint(x: lx + 32, y: ly + 11), CGPoint(x: lx + 34, y: ly + 22), CGPoint(x: lx + 31, y: ly + 36)),
                (CGPoint(x: lx + 40, y: ly + 12), CGPoint(x: lx + 42, y: ly + 24), CGPoint(x: lx + 39, y: ly + 35)),
            ]
            for (start, ctrl, end) in stripes {
                ctx.beginPath()
                ctx.move(to: start)
                ctx.addQuadCurve(to: end, control: ctrl)
                ctx.strokePath()
            }

            // --- Ears ---
            func drawEllipse(cx: CGFloat, cy: CGFloat, rx: CGFloat, ry: CGFloat, rotation: CGFloat, color: UIColor, startAngle: CGFloat = 0, endAngle: CGFloat = .pi * 2) {
                ctx.saveGState()
                ctx.translateBy(x: cx, y: cy)
                ctx.rotate(by: rotation)
                ctx.setFillColor(color.cgColor)
                let rect = CGRect(x: -rx, y: -ry, width: rx * 2, height: ry * 2)
                if startAngle == 0 && endAngle == .pi * 2 {
                    ctx.fillEllipse(in: rect)
                } else {
                    // For half ellipse (ear backs): scale a circle arc
                    ctx.scaleBy(x: 1, y: ry / rx)
                    ctx.beginPath()
                    ctx.addArc(center: .zero, radius: rx, startAngle: startAngle, endAngle: endAngle, clockwise: false)
                    ctx.closePath()
                    ctx.fillPath()
                }
                ctx.restoreGState()
            }

            // Left ear
            drawEllipse(cx: lx + 43, cy: ly + 3, rx: 5, ry: 6, rotation: -0.15,
                        color: UIColor(red: 1, green: 0.549, blue: 0, alpha: 1))
            // Right ear
            drawEllipse(cx: lx + 57, cy: ly + 3, rx: 5, ry: 6, rotation: 0.15,
                        color: UIColor(red: 1, green: 0.549, blue: 0, alpha: 1))

            // Black ear backs (half ellipse, PI to 2*PI in canvas = top half in flipped coords)
            drawEllipse(cx: lx + 43, cy: ly + 1, rx: 4, ry: 4, rotation: -0.15,
                        color: UIColor(red: 0.102, green: 0.102, blue: 0.102, alpha: 1),
                        startAngle: .pi, endAngle: .pi * 2)
            drawEllipse(cx: lx + 57, cy: ly + 1, rx: 4, ry: 4, rotation: 0.15,
                        color: UIColor(red: 0.102, green: 0.102, blue: 0.102, alpha: 1),
                        startAngle: .pi, endAngle: .pi * 2)

            // White ear spots
            ctx.setFillColor(UIColor.white.cgColor)
            ctx.fillEllipse(in: CGRect(x: lx + 43 - 2, y: ly + 1 - 2, width: 4, height: 4))
            ctx.fillEllipse(in: CGRect(x: lx + 57 - 2, y: ly + 1 - 2, width: 4, height: 4))

            // --- Head ---
            ctx.setFillColor(UIColor(red: 1, green: 0.647, blue: 0, alpha: 1).cgColor) // #FFA500
            ctx.fillEllipse(in: CGRect(x: lx + 50 - 13, y: ly + 13 - 12, width: 26, height: 24))

            // White muzzle
            ctx.setFillColor(UIColor(red: 1, green: 0.973, blue: 0.882, alpha: 1).cgColor) // #FFF8E1
            ctx.fillEllipse(in: CGRect(x: lx + 50 - 8, y: ly + 18 - 6, width: 16, height: 12))

            // White cheek patches
            ctx.setFillColor(UIColor(red: 1, green: 0.953, blue: 0.878, alpha: 1).cgColor)
            drawEllipse(cx: lx + 42, cy: ly + 15, rx: 4, ry: 3, rotation: -0.3,
                        color: UIColor(red: 1, green: 0.953, blue: 0.878, alpha: 1))
            drawEllipse(cx: lx + 58, cy: ly + 15, rx: 4, ry: 3, rotation: 0.3,
                        color: UIColor(red: 1, green: 0.953, blue: 0.878, alpha: 1))

            // Face stripes
            ctx.setStrokeColor(UIColor(red: 0.102, green: 0.102, blue: 0.102, alpha: 1).cgColor)
            ctx.setLineWidth(2)
            let faceStripes: [(CGPoint, CGPoint, CGPoint)] = [
                (CGPoint(x: lx + 40, y: ly + 6), CGPoint(x: lx + 42, y: ly + 9), CGPoint(x: lx + 44, y: ly + 8)),
                (CGPoint(x: lx + 38, y: ly + 9), CGPoint(x: lx + 41, y: ly + 12), CGPoint(x: lx + 43, y: ly + 11)),
                (CGPoint(x: lx + 60, y: ly + 6), CGPoint(x: lx + 58, y: ly + 9), CGPoint(x: lx + 56, y: ly + 8)),
                (CGPoint(x: lx + 62, y: ly + 9), CGPoint(x: lx + 59, y: ly + 12), CGPoint(x: lx + 57, y: ly + 11)),
            ]
            for (start, ctrl, end) in faceStripes {
                ctx.beginPath()
                ctx.move(to: start)
                ctx.addQuadCurve(to: end, control: ctrl)
                ctx.strokePath()
            }

            // Eyes - white
            ctx.setFillColor(UIColor.white.cgColor)
            drawEllipse(cx: lx + 45, cy: ly + 11, rx: 3.5, ry: 2.5, rotation: -0.1, color: .white)
            drawEllipse(cx: lx + 55, cy: ly + 11, rx: 3.5, ry: 2.5, rotation: 0.1, color: .white)

            // Amber iris
            ctx.setFillColor(UIColor(red: 0.902, green: 0.659, blue: 0, alpha: 1).cgColor) // #E6A800
            ctx.fillEllipse(in: CGRect(x: lx + 46 - 2, y: ly + 11 - 2, width: 4, height: 4))
            ctx.fillEllipse(in: CGRect(x: lx + 56 - 2, y: ly + 11 - 2, width: 4, height: 4))

            // Black pupils
            ctx.setFillColor(UIColor.black.cgColor)
            ctx.fillEllipse(in: CGRect(x: lx + 46 - 1, y: ly + 11 - 1, width: 2, height: 2))
            ctx.fillEllipse(in: CGRect(x: lx + 56 - 1, y: ly + 11 - 1, width: 2, height: 2))

            // Nose
            ctx.setFillColor(UIColor(red: 0.831, green: 0.341, blue: 0.42, alpha: 1).cgColor) // #D4576B
            ctx.fillEllipse(in: CGRect(x: lx + 50 - 3, y: ly + 16 - 2, width: 6, height: 4))
            ctx.setFillColor(UIColor(red: 0.102, green: 0.102, blue: 0.102, alpha: 1).cgColor)
            // Bottom half of nose
            ctx.saveGState()
            ctx.translateBy(x: lx + 50, y: ly + 15.5)
            ctx.scaleBy(x: 1, y: 1.5 / 2.5)
            ctx.beginPath()
            ctx.addArc(center: .zero, radius: 2.5, startAngle: 0, endAngle: .pi, clockwise: false)
            ctx.closePath()
            ctx.fillPath()
            ctx.restoreGState()

            // Mouth line
            ctx.setStrokeColor(UIColor(red: 0.102, green: 0.102, blue: 0.102, alpha: 1).cgColor)
            ctx.setLineWidth(1)
            ctx.beginPath()
            ctx.move(to: CGPoint(x: lx + 50, y: ly + 17.5))
            ctx.addLine(to: CGPoint(x: lx + 50, y: ly + 19))
            ctx.strokePath()
            // Mouth arcs
            ctx.beginPath()
            ctx.addArc(center: CGPoint(x: lx + 47, y: ly + 19), radius: 3,
                       startAngle: 0, endAngle: .pi, clockwise: false)
            ctx.strokePath()
            ctx.beginPath()
            ctx.addArc(center: CGPoint(x: lx + 53, y: ly + 19), radius: 3,
                       startAngle: 0, endAngle: .pi, clockwise: false)
            ctx.strokePath()

            // Whiskers
            ctx.setStrokeColor(UIColor.white.cgColor)
            ctx.setLineWidth(1.5)
            ctx.setLineCap(.round)
            let whiskers: [(CGPoint, CGPoint)] = [
                (CGPoint(x: lx + 43, y: ly + 17), CGPoint(x: lx + 32, y: ly + 15)),
                (CGPoint(x: lx + 43, y: ly + 18), CGPoint(x: lx + 31, y: ly + 19)),
                (CGPoint(x: lx + 43, y: ly + 19), CGPoint(x: lx + 33, y: ly + 23)),
                (CGPoint(x: lx + 57, y: ly + 17), CGPoint(x: lx + 68, y: ly + 15)),
                (CGPoint(x: lx + 57, y: ly + 18), CGPoint(x: lx + 69, y: ly + 19)),
                (CGPoint(x: lx + 57, y: ly + 19), CGPoint(x: lx + 67, y: ly + 23)),
            ]
            for (from, to) in whiskers {
                ctx.beginPath()
                ctx.move(to: from)
                ctx.addLine(to: to)
                ctx.strokePath()
            }

            // --- Legs ---
            ctx.setStrokeColor(UIColor(red: 1, green: 0.549, blue: 0, alpha: 1).cgColor)
            ctx.setLineWidth(4)
            ctx.setLineCap(.round)

            if isOnGround && isPlaying {
                let legPhase = sin(CGFloat(frameCount) * 0.3)
                let legs: [(CGPoint, CGPoint)] = [
                    (CGPoint(x: lx + 38, y: ly + 40), CGPoint(x: lx + 38 + legPhase * 6, y: ly + 50)),
                    (CGPoint(x: lx + 44, y: ly + 40), CGPoint(x: lx + 44 - legPhase * 6, y: ly + 50)),
                    (CGPoint(x: lx + 16, y: ly + 40), CGPoint(x: lx + 16 - legPhase * 6, y: ly + 50)),
                    (CGPoint(x: lx + 22, y: ly + 40), CGPoint(x: lx + 22 + legPhase * 6, y: ly + 50)),
                ]
                for (from, to) in legs {
                    ctx.beginPath()
                    ctx.move(to: from)
                    ctx.addLine(to: to)
                    ctx.strokePath()
                }
            } else if !isOnGround {
                let legs: [(CGPoint, CGPoint)] = [
                    (CGPoint(x: lx + 38, y: ly + 40), CGPoint(x: lx + 42, y: ly + 46)),
                    (CGPoint(x: lx + 44, y: ly + 40), CGPoint(x: lx + 48, y: ly + 46)),
                    (CGPoint(x: lx + 16, y: ly + 40), CGPoint(x: lx + 12, y: ly + 46)),
                    (CGPoint(x: lx + 22, y: ly + 40), CGPoint(x: lx + 18, y: ly + 46)),
                ]
                for (from, to) in legs {
                    ctx.beginPath()
                    ctx.move(to: from)
                    ctx.addLine(to: to)
                    ctx.strokePath()
                }
            } else {
                let legs: [(CGPoint, CGPoint)] = [
                    (CGPoint(x: lx + 38, y: ly + 40), CGPoint(x: lx + 38, y: ly + 50)),
                    (CGPoint(x: lx + 44, y: ly + 40), CGPoint(x: lx + 44, y: ly + 50)),
                    (CGPoint(x: lx + 16, y: ly + 40), CGPoint(x: lx + 16, y: ly + 50)),
                    (CGPoint(x: lx + 22, y: ly + 40), CGPoint(x: lx + 22, y: ly + 50)),
                ]
                for (from, to) in legs {
                    ctx.beginPath()
                    ctx.move(to: from)
                    ctx.addLine(to: to)
                    ctx.strokePath()
                }
            }
        }
        return SKTexture(image: image)
    }
}
